### PR TITLE
feat: add cypress automated UI testing framework

### LINKED ISSUE #476: Shopping App Website: UI testing
Automated UI testing (https://webdriver.io/ or https://www.cypress.io).

**Acceptance Criteria**

All aspects of the UI interaction and behavior should be covered by a UI testing framework.

### PR DESCRIPTION
**Pull request by mrmodise**
Closes #476

### CODE REVIEW

File: packages/shopping/cypress/integration/cart.spec.js
Comment (dougal83) [inline packages/shopping/cypress/integration/cart.spec.js:23]: We really do not want `.wait()` if we can avoid it.  Please try something like:
````
        cy
          .get('.list-group')
          .find('li')
          .get('input[name="cartItem"]')
          .should('be.visible')
          .check(),
````
Good article here: https://cypress.io/blog/2019/01/22/when-can-the-test-click/

File: packages/shopping/cypress/integration/cart.spec.js
Comment (dougal83) [inline packages/shopping/cypress/integration/cart.spec.js:36]: Again. Try to remove wait().

File: packages/shopping/cypress/integration/cart.spec.js
Comment (dougal83) [inline packages/shopping/cypress/integration/cart.spec.js:49]: Again. Try to remove wait().

**Review by dougal83 — CHANGES_REQUESTED** [review]
To code owners: With regards to folder structure, would calling the folder 'e2e' or even '__tests__/e2e' instead of 'cypress' might be a step in the right direction?

Also inline with the project being a typescript project, would it be best to use `.ts` files?

File: packages/shopping/cypress/integration/cart.spec.js
Comment (mrmodise) [reply packages/shopping/cypress/integration/cart.spec.js:23]: Okay cool will update. Thanks

**Comment by mrmodise** [comment]
> To code owners: With regards to folder structure, would calling the folder `e2e` or even `__tests__/e2e` instead of `cypress` might be a step in the right direction?
> 
> Also inline with the project being a typescript project, would it be best to use `.ts` files?

@dougal83 I updated the tests to be under __tests__. I will explore the option of having the tests running on ts too

**Comment by mrmodise** [comment]
> To code owners: With regards to folder structure, would calling the folder `e2e` or even `__tests__/e2e` instead of `cypress` might be a step in the right direction?
> 
> Also inline with the project being a typescript project, would it be best to use `.ts` files?

@hacksparrow @jannyHou please advise

**Comment by dhmlau** [comment]
I'm not the codeowner :) but +1 on keeping the naming consistent for `e2e`, just like we use mocha for non-UI tests but still doesn't use the test framework name as the folder name.

@nabdelgadir, I believe you have some experience in cypress. Would you like to review this PR as well? Thanks.

**Review by nabdelgadir — COMMENTED** [review]
Thanks for the PR! I'll review it tomorrow, but in the meantime, can you add the copyright header to the files? Thanks!

```
// Copyright IBM Corp. 2020. All Rights Reserved.
// Node module: loopback4-example-shopping
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
```

**Comment by dhmlau** [comment]
Thanks @nabdelgadir. We have a tool to add the copyright headers, so we can do it in a separate PR.

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (nabdelgadir) [inline packages/shopping/src/__tests__/e2e/integration/cart.spec.js:9]: When running this locally, it failed because it couldn't click on this:
Screen Shot 2020-10-21 at 9 20 53 AM

Can you verify the login modal isn't visible before this line?

File: packages/shopping/src/__tests__/e2e/integration/profile.spec.js
Comment (nabdelgadir) [inline packages/shopping/src/__tests__/e2e/integration/profile.spec.js:20]: I'm not sure why this test is failing, but it might be a timing issue. Can you confirm if it works properly for you?

Screen Shot 2020-10-21 at 9 27 03 AM

File: packages/shopping/src/__tests__/e2e/integration/profile.spec.js
Comment (nabdelgadir) [inline packages/shopping/src/__tests__/e2e/integration/profile.spec.js:15]: Minor nitpick: Loopback -> LoopBack

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (nabdelgadir) [inline packages/shopping/src/__tests__/e2e/integration/cart.spec.js:35]: Similarly here, the modal from the previous test is still open so this test fails:

Screen Shot 2020-10-21 at 9 24 25 AM

**Review by nabdelgadir — COMMENTED** [review]
Nice work! A few comments:

1) For the selectors, Cypress recommends using `data-*` attributes instead of ids and classes so they're not affected by CSS/JS changes: https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements

2) Can you add some documentation so users know how to run the tests (e.g. starting the server first)?

3) For the code owners, do we want to run this on CI?

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:9]: Will check to ensure its visible before executing the click

File: packages/shopping/src/__tests__/e2e/integration/profile.spec.js
Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/profile.spec.js:20]: Probably same issue will add `.should('be.visible')` to ensure it always wait for the item to display

File: packages/shopping/src/__tests__/e2e/integration/profile.spec.js
Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/profile.spec.js:15]: nice catch! 😆

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:35]: Will apply same pattern, as above. Interesting enough on my machine they're all working

Comment (dougal83) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:35]: > Will apply same pattern, as above. Interesting enough on my machine they're all working

Worked locally for me too.  We're all learning. Great work btw.

Comment (nabdelgadir) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:35]: Did you run only this test or all the tests? By itself, the test runs fine, but the failure could just be from my machine.

> Great work btw.

+1

Comment (dougal83) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:35]: I ran all the tests and it went smoothly at the time of review. For the best that you had an issue as it prompted further refinement. 👍

**Comment by raymondfeng** [comment]
@mrmodise Thank you for the contribution. Please resolve the conflicts before we can land it.

BTW, you can run `lb4 copyright` to fix the file headers.

**Comment by mrmodise** [comment]
> @mrmodise Thank you for the contribution. Please resolve the conflicts before we can land it.
> 
> BTW, you can run `lb4 copyright` to fix the file headers.

Thanks applied the change

**Comment by mrmodise** [comment]
> Nice work! A few comments:
> 
> 1. For the selectors, Cypress recommends using `data-*` attributes instead of ids and classes so they're not affected by CSS/JS changes: https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements
> 2. Can you add some documentation so users know how to run the tests (e.g. starting the server first)?
> 3. For the code owners, do we want to run this on CI?

1. For the selectors, Cypress recommends using `data-*` attributes instead of ids and classes so they're not affected by CSS/JS changes: https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements [x]
2. Can you add some documentation so users know how to run the tests (e.g. starting the server first)? [x]

File: README.md
Comment (nabdelgadir) [inline README.md:211]:
```suggestion
$ npm test
```

File: README.md
Comment (nabdelgadir) [inline README.md:197]: I see you have this under the Authentication section, but can you move it under the usage section?

File: README.md
Comment (nabdelgadir) [inline README.md:202]: Can you also add a note for them to make sure the server is running?

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (nabdelgadir) [inline packages/shopping/src/__tests__/e2e/integration/cart.spec.js:20]: I can't reply to the old thread for some reason, but this test is still failing for me. The login modal is still open, so despite the link being visible, it can't be clicked on:

image

I think you'll also have to verify that the modal not visible.

**Review by nabdelgadir — COMMENTED** [review]
Thanks for applying the feedback! I have a few more comments 😄

**Comment by dhmlau** [comment]
> For the code owners, do we want to run this on CI?
I don't think we need to run this on CI. This is to provide an example on how users can add their automated UI testing.

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:20]: Difficult to fix this 1 on my end since the test suite is passing. How are you executing these?

File: README.md
Comment (mrmodise) [reply README.md:202]: Added

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (nabdelgadir) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:20]: I'm using the "Run all specs" option

File: README.md
Comment (mrmodise) [reply README.md:197]: Updated

File: packages/shopping/src/__tests__/e2e/integration/cart.spec.js
Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:20]: Screenshot 2020-10-22 at 17 01 45

Comment (mrmodise) [reply packages/shopping/src/__tests__/e2e/integration/cart.spec.js:20]: Screenshot 2020-10-22 at 17 07 29

File: README.md
Comment (mrmodise) [reply README.md:211]: Applied change

File: .eslintrc.js
Comment (nabdelgadir) [inline .eslintrc.js:14]: Is this intentional?

**Review by nabdelgadir — APPROVED** [review]
💯 great stuff!

File: .eslintrc.js
Comment (mrmodise) [reply .eslintrc.js:14]: Correct. express-eslint seems to trigger floating promises linting issues on JQuery UI files

Comment (dougal83) [reply .eslintrc.js:14]: Would it be better to add `// eslint-disable-next-line @typescript-eslint/no-floating-promises` to `public/ui.js:L10` since it is just the one error?

**Review by dougal83 — APPROVED** [review]
Looks good.  Maybe recreate `packages/shopping/package-lock.json` as there appear to be some duplicate lines (delete, npm i)?

File: .eslintrc.js
Comment (mrmodise) [reply .eslintrc.js:14]: Let me have a look

Comment (mrmodise) [reply .eslintrc.js:14]: Works will update

**Comment by mrmodise** [comment]
> Looks good. Maybe recreate `packages/shopping/package-lock.json` as there appear to be some duplicate lines (delete, npm i)?

nice catch! updated. Thanks

**Comment by nabdelgadir** [comment]
Thanks for your contribution @mrmodise! 🎉

**Pull request merged**
